django/applications/catmaid/static/js/image-block.js

```diff
@@ -0,0 +1,63 @@
+(function (CATMAID) {
+
+  "use strict";
+
+  var ImageBlock = {};
+
+  ImageBlock.Cache = class Cache {
+    constructor(source, capacity = 256) {
+      this.source = source;
+      this._cache = new CATMAID.LRUCache(capacity);
+      this._stateIDs = new Map();
+      this._dirty = new Set();
+
+      this._cache.on(CATMAID.LRUCache.EVENT_EVICTED, this._onBlockEviction, this);
+      CATMAID.asEventSource(this);
+    }
+
+    readBlock(...zoomBlockCoord) {
+      let blockKey = zoomBlockCoord.join('/');
+      let block = this._cache.get(blockKey);
+
+      if (block) {
+        return Promise.resolve(block);
+      }
+
+      return this.source.readBlock(...zoomBlockCoord)
+          .then(block => {
+            this._stateIDs.set(blockKey, 'fresh');  // TODO
+            this._cache.set(blockKey, block);
+            return block;
+          });
+    }
+
+    setBlock(zoomLevel, x, y, z, block) {
+      let blockKey = [zoomLevel, x, y, z].join('/');
+
+      this._dirty.add(blockKey);
+      this._cache.set(blockKey, block);
+
+      this.trigger(
+          ImageBlock.Cache.EVENT_BLOCK_CHANGED,
+          {zoomLevel, x, y, z, block});
+    }
+
+    _onBlockEviction(key, block) {
+      let stateID = this._stateIDs.get(key);
+      this._stateIDs.delete(key);
+
+      if (this._dirty.delete(key)) {
+        let [zoomLevel, x, y, z] = key.split('/');
+        this.trigger(
+            ImageBlock.Cache.EVENT_DIRTY_BLOCK_EVICTED,
+            {zoomLevel, x, y, z, stateID, block});
```

Comment:
That confused me for a second, but it indeed works and I was just not aware of these ES6 object literals---makes it certainly more readable.